android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt

```diff
@@ -282,13 +312,43 @@ constructor(
 
       softDeleteResources(questionnaireConfig)
 
+      retireUsedQuestionnaireUniqueId(questionnaireConfig, currentQuestionnaireResponse)
+
       val idTypes =
         bundle.entry?.map { IdType(it.resource.resourceType.name, it.resource.logicalId) }
           ?: emptyList()
       onSuccessfulSubmission(idTypes, currentQuestionnaireResponse)
     }
   }
 
+  suspend fun retireUsedQuestionnaireUniqueId(
+    questionnaireConfig: QuestionnaireConfig,
+    questionnaireResponse: QuestionnaireResponse
+  ) {
+    if (uniqueIdResourcePair != null && questionnaireConfig.uniqueIdAssignment != null) {
+      val (id, resource) = uniqueIdResourcePair!!
+      val uniqueIdLinkId = questionnaireConfig.uniqueIdAssignment!!.linkId
+      val submittedUniqueId =
+        questionnaireResponse.find(uniqueIdLinkId)?.answer?.first()?.value.toString()
+
+      // Update Group resource. Can be extended in future to support other resources
+      if (resource is Group) {
+        resource.characteristic.onEach {
+          if (it.hasValueCodeableConcept() && it.valueCodeableConcept.text == submittedUniqueId) {
+            it.exclude = true
+            return@onEach
+          }
+        }
+
+        // Mark Group as inactive when all IDs are retired
+        resource.active = resource.characteristic.any { !it.exclude }
```

Comment:
I'm worried about the performance implications of doing 2 full scans of the characteristic list.

If we do not allow the ID FHIRPath expression to be configurable, we can use a counter to do this and have everything in constant time. We can set `group.quantity` to the index of the next unused ID, then all we do here is say `resource.quantity++`. Then set to readonly if `resource.quantity > resource.characteristic.size`

Reply:
The recommended approach will work if we hard-code this functionality for a Group resource. However, the [Group.quantity](http://hl7.org/fhir/R4B/group-definitions.html#Group.quantity) property is supposed to be used as the count of the members according to the FHIR documentation.

Reply:
A boolean state can be updated in the first iteration of the characteristics to be used as the value for `resource.active` to prevent the second scan.

Reply:
Yea true and true, but we only support group now so we don't lose anything by making this only work for group, since we're already not following the semantics of characteristic, seems fine to not follow that for quantity

Reply:
But this would be a meaningful change, late in the process, for efficiency reasons I'm ok w/keeping this as is if we've run some tests with a Group of reasonable size on low end devices and this has negligible/no impact on the performance

Reply:
@allan-on Can you verify the performance impact? Do you know on average how many items we would expect in the `Group.characteristics` property?